content/5_Plat/Bin_Jump.mdx

```diff
@@ -353,6 +353,147 @@ int main() {
 
 <IncompleteSection />
 
+### Solution (Company Queries II)
+
+First, let $\texttt{st}, \texttt{en}, \texttt{up}$ be the time-in, time-out, and ancestor table for all nodes in the tree. 
+
+These can be filled with a DFS traversal. $\texttt{up}$ can be generated because depth first allows the ancestors to be filled first before traversing the current node.
+
+With this information, we can declare node $a$ an ancestor of $b$ if $$\texttt{st}[a] \le \texttt{st}[b] \&\& \texttt{en}[a] \ge \texttt{en}[b]$$
+
+We know that if $a$ is an ancestor of $b$ or $b$ is an ancestor of $a$, the answer will be $a$ and $b$ respectively. Otherwise, we can loop from the largest jumps to the the direct ancestor of $a$ to find the LCA of $a$ and $b$. If $\texttt{up}[a][i]$ is not an ancestor of $b$, then we can set $a$ to be $\texttt{up}[a][i]$, else, we will decrement $i$ and try to find a lower common ancestor. Afterwards, our answer is not $a$, but the parent of $a$. 
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+CodeSnip{Benq Template}
+	
+int N, Q, T = 1;
+vi st, en;
+int up[200005][20];
+vi adj[200005];
+
+void dfs(int v, int p) {
+	st[v] = T++;
+	up[v][0] = p;
+
+	FOR(i, 1, 20) {
+		up[v][i] = up[up[v][i-1]][i-1];
+	}
+
+	each(x, adj[v]) if(x != p) dfs(x, v);
+
+	en[v] = T-1;
+}
+
+bool anc(int a, int b) {
+	return st[a] <= st[b] && en[a] >= en[b];
+}
+
+int LCA(int a, int b) {
+	for (int i = 19; i >= 0; --i) {
+		if (!anc(up[a][i], b))
+			a = up[a][i];
+	}
+	return up[a][0];
+}
+	
+int main() {   
+    setIO("");
```

Comment:
:( still spaces here

Reply:
wdym?

Reply:
for some reason the commit you made got reverted here https://github.com/cpinitiative/usaco-guide/pull/953/commits/7949002029b87999af804d2dc5baa9a2ad33c7b7

Reply:
huh? im not seeing any reverted changes

Reply:
reload?

Reply:
actually i'll just commit it